src/controller/BUILD.gn

```diff
@@ -24,41 +24,91 @@ source_set("nodatamodel") {
   public_deps = [ "${chip_root}/src/app" ]
 }
 
+CHIP_CONTROLLER_SOURCES = [
+  "AbstractDnssdDiscoveryController.cpp",
+  "AutoCommissioner.cpp",
+  "CHIPCommissionableNodeController.cpp",
+  "CHIPDeviceControllerFactory.cpp",
+  "CHIPDeviceControllerFactory.h",
+  "CommissioneeDeviceProxy.cpp",
+  "CommissionerDiscoveryController.cpp",
+  "CommissionerDiscoveryController.h",
+  "CommissioningDelegate.cpp",
+  "ExampleOperationalCredentialsIssuer.cpp",
+  "ExampleOperationalCredentialsIssuer.h",
+  "SetUpCodePairer.cpp",
+]
+
+CHIP_READ_CLIENT_SOURCES = [
+  "CHIPDeviceController.cpp",
+  "CommissioningWindowOpener.cpp",
+  "CommissioningWindowOpener.h",
+  "CurrentFabricRemover.cpp",
+  "CurrentFabricRemover.h",
+]
+
+if (!chip_controller || !chip_build_controller) {
+  # This source set exists specifically to deny including them without dependencies
+  # if "controller" sources do not contain them
+  #
+  # They are NOT intended to be used directly
+  #
+  # Also note: these source_sets are NOT compilable/usable.
+  #            expect tools like clang-tidy to fail to find include paths if these
+  #            are enabled.
+  source_set("gen_check_chip_controller_sources") {
```

Comment:
These source sets could be defined always and only pulled as a depedency of ":controller" if needed, right? Then you wouldn't need to introduce a new pattern of defining source lists as regular variables.

Reply:
That is correct. The bit that I am trying to avoid is if controller is built without `chip_controller` or `chip_build_controller` if GN sees no build rules it will just allow these files to be built.

In practice, assume I do not build chip_controller, then a file could be created `x.cpp` with `#include <controller/CommissioneeDeviceProxy.h>` and it will compile. With this change gn will complain of `include not allowed, you need a dependency of gen_check_chip_controller_sources` because now gn sees at least one source set that contains this file.

Overall I need GN to actually be aware that it needs to manage each file. The conditionals here subvert this.

Reply:
I also think this is what should be done. #28149 seems to just added defines on the source set itself (which I am unclear why it would work ... planning to revert that).

We should guard everything via gn. The somewhat harder part seems to be that src/app is not managed via GN at all (and then we have cmake rules trying to cover things and ensure that what cmake brings in is also available from gn).

Reply:
I don't think I can do public_deps because they are not inherited, so I need everything that needs the headers depend on the `dummy_module` for the header include. `ExampleOperationalCredentialsIssuer.h` for example is used in many places.

I would like a followup to split these into real stand-alone modules.

Reply:
Right, I should probably review the CMake integration with regards to that as some parts like `app/server` are now built twice. Don't remember how we got there, but initially there was a cleaner separation between app and `libCHIP` and that's why app was pulled in by CMake directly. But it may no longer make sense and it may be better to just pass ZAP information to GN and build everything there.